### ISSUE TITLE
CORS preflight request with invalid origin is not handled properly when custom origin check is set

### ISSUE THREAD

**Issue by andrejleitner** [socketio/socket.io#3947]
**Describe the bug**
CORS preflight request with invalid origin is handled differently if the origin is verified with a custom function. It seems that the processing of the request continues with standard engine.io verification and results in `Bad handshake method` error (since only `GET` is allowed in the handshake).

**To Reproduce**

Socket.IO server version: `4.1.2`

**Origin defined with custom function**

_Server_

```js
function verifyOrigin(origin, callback) {
    if (origin === 'http://origin.ok') {
        callback(null, true);
    } else {
        callback(new Error('Access not allowed from the specified origin: ' + origin), false);
    }
}

const options = {
    cors: {
        credentials: true,
        methods: ['GET', 'POST'],
        origin: verifyOrigin
    }
};
const io = require('socket.io')(options);

io.on("connection", socket => {
    console.log('Socket connected.')
});

io.listen(3000);
```

_CORS preflight imitation_

```
// valid origin
curl -i -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://origin.ok" 'http://localhost:3000/socket.io/?EIO=4&transport=polling'
HTTP/1.1 204 No Content
Access-Control-Allow-Origin: http://origin.ok
Vary: Origin, Access-Control-Request-Headers
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET,POST
Content-Length: 0
Date: Sat, 29 May 2021 08:28:32 GMT
Connection: keep-alive
Keep-Alive: timeout=5

// invalid origin
curl -i -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://origin.wrong" 'http://localhost:3000/socket.io/?EIO=4&transport=polling'
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Sat, 29 May 2021 08:28:42 GMT
Connection: keep-alive
Keep-Alive: timeout=5
Transfer-Encoding: chunked

{"code":2,"message":"Bad handshake method"}
```

**Origin defined statically**

_Server_

```js
const options = {
    cors: {
        credentials: true,
        methods: ['GET', 'POST'],
        origin: 'http://origin.ok'
    }
};
const io = require('socket.io')(options);

io.on("connection", socket => {
    console.log('Socket connected.')
});

io.listen(3000);
```

_CORS preflight imitation_

```
// valid origin
curl -i -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://origin.ok" 'http://localhost:3000/socket.io/?EIO=4&transport=polling'
HTTP/1.1 204 No Content
Access-Control-Allow-Origin: http://origin.ok
Vary: Origin, Access-Control-Request-Headers
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET,POST
Content-Length: 0
Date: Sat, 29 May 2021 09:13:50 GMT
Connection: keep-alive
Keep-Alive: timeout=5

// invalid origin
curl -i -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://origin.wrong" 'http://localhost:3000/socket.io/?EIO=4&transport=polling'
HTTP/1.1 204 No Content
Access-Control-Allow-Origin: http://origin.ok
Vary: Origin, Access-Control-Request-Headers
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET,POST
Content-Length: 0
Date: Sat, 29 May 2021 09:13:55 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

**Expected behavior**
Correct response to the CORS preflight request in both cases.

**Platform:**
 - Node: v14.17.0

**Comment by darrachequesne** [comment]
If I understand correctly, this behavior comes from the `cors` package:

Here: https://github.com/expressjs/cors/blob/c49ca10e92ac07f98a3b06783d3e6ba0ea5b70c7/lib/index.js#L219-L226

```js
originCallback(req.headers.origin, function (err2, origin) {
  if (err2 || !origin) {
    next(err2);
  } else {
    corsOptions.origin = origin;
    cors(corsOptions, req, res, next);
  }
});
```

- with a plain string, the callback is always called with `null`, and the handler returns an HTTP 204 response ([here](https://github.com/expressjs/cors/blob/c49ca10e92ac07f98a3b06783d3e6ba0ea5b70c7/lib/index.js#L178-L180))
- with a function, the `next` handler is called directly, leading to the HTTP 400 error at the Engine.IO level

We could handle the error [here](https://github.com/socketio/engine.io/blob/7e9e544a9ac4365ef055c9482cc1e9fed624031c/lib/server.js#L254-L256), or fix the issue (if it's actually an issue) in the `cors` package directly. @andrejleitner what is your opinion?

**Comment by andrejleitner** [comment]
Thanks @darrachequesne for looking into it. I think you are right, I debugged it to the same result. Atm, I don't know what's the right place where to handle it. If you are in contact with someone from `cors` package maybe you can discuss it, but definitely it should not "leak" down into Engine.IO `verify` method, right (preflight request ending on handshake method check)? At the end, the response error is misleading.

**Comment by darrachequesne** [comment]
@andrejleitner here we go: https://github.com/expressjs/cors/issues/241

**Comment by andrejleitner** [comment]
Hey @darrachequesne, there seems to be no activity on that issue 😞.

**Comment by darrachequesne** [comment]
@andrejleitner well, it's open source, 13 days are not that much :smile: 

Would you have time to open a pull request on the `cors` repository? Maybe it will be easier for the maintainer to acknowledge the issue if we already have a fix.

**Comment by andrejleitner** [comment]
Resolved in the discussion on `cors` side.

TLDR;

Invalid origin shoud be rejected by calling `callback(null,[])`, e.g.:

``` js
if (origin === 'http://origin.ok') {
        callback(null, true);
    } else {
        callback(null, []);
    }
```
